File: app/src/main/java/com/maybescanner/PhaseResult.java
package com.maybescanner;

import org.json.JSONObject;

import java.util.Locale;

/** Per-phase probe outcome aligned with sidecar PhaseResult v1. */
final class PhaseResult {
    final String phase;
    final String status;
    final long durationMs;
    final String errorCode;
    final boolean retryable;

    PhaseResult(String phase, String status, long durationMs, String errorCode, boolean retryable) {
        this.phase = phase == null ? "" : phase;
        this.status = status == null ? "failed" : status;
        this.durationMs = Math.max(0L, durationMs);
        this.errorCode = errorCode == null ? "" : errorCode;
        this.retryable = retryable;
    }

    static PhaseResult success(String phase, long durationMs) {
        return new PhaseResult(phase, "success", durationMs, "", false);
    }

    static PhaseResult failure(String phase, long durationMs, Exception error, String errorCode) {
        String code = errorCode == null || errorCode.isEmpty() ? classifyCode(error, phase) : errorCode;
        return new PhaseResult(phase, statusFromCode(code), durationMs, code, retryableFromCode(code));
    }

    JSONObject toJson() throws Exception {
        JSONObject payload = new JSONObject();
        payload.put("phase", phase);
        payload.put("status", status);
        payload.put("duration_ms", durationMs);
        payload.put("retryable", retryable);
        if (!errorCode.isEmpty()) payload.put("error_code", errorCode);
        payload.put("evidence", new JSONObject());
        return payload;
    }

    static String classifyCode(Exception error, String phase) {
        String prefix = "SCAN";
        if ("dns".equals(phase)) {
            prefix = "DNS";
        } else if ("tcp".equals(phase)) {
            prefix = "TCP_CONNECT";
        } else if ("tls".equals(phase)) {
            prefix = "TLS_HANDSHAKE";
        } else if ("http1".equals(phase) || "http2".equals(phase) || "http".equals(phase)) {
            prefix = "HTTP";
        } else if ("route".equals(phase)) {
            prefix = "ROUTE";
        }
        if (error == null) {
            return prefix + "_FAILED";
        }
        String lower = String.valueOf(error.getMessage()).toLowerCase(Locale.US);
        if (lower.contains("timeout") || lower.contains("timed out")) {
            return prefix + "_TIMEOUT";
        }
        if (lower.contains("reset")) {
            return prefix + "_RESET";
        }
        if (lower.contains("refused")) {
            return prefix + "_REFUSED";
        }
        return prefix + "_FAILED";
    }

    static String statusFromCode(String code) {
        String upper = code == null ? "" : code.toUpperCase(Locale.US);
        if (upper.isEmpty()) {
            return "failed";
        }
        if (upper.endsWith("_TIMEOUT")) {
            return "timeout";
        }
        if (upper.endsWith("_REFUSED")) {
            return "refused";
        }
        if (upper.endsWith("_RESET")) {
            return "reset";
        }
        if (upper.contains("_SKIPPED") || upper.endsWith("_EXCLUDED")) {
            return "skipped";
        }
        if (upper.contains("_UNSUPPORTED")) {
            return "unsupported";
        }
        return "failed";
    }

    static boolean retryableFromCode(String code) {
        String upper = code == null ? "" : code.toUpperCase(Locale.US);
        return upper.endsWith("_TIMEOUT") || upper.endsWith("_RESET");
    }

    static String httpPhaseFromAlpn(String alpn) {
        if (alpn != null && alpn.toLowerCase(Locale.US).contains("h2")) {
            return "http2";
        }
        return "http1";
    }

    /** User-visible label derived from stable error_code, not probe exception text. */
    static String displayLabel(String errorCode, String phase, String status) {
        if (errorCode != null && !errorCode.isEmpty()) {
            return humanizeCode(errorCode);
        }
        if ("success".equals(status)) {
            return capitalize(phase) + " ok";
        }
        if (phase != null && !phase.isEmpty()) {
            return capitalize(phase) + " " + (status == null || status.isEmpty() ? "failed" : status);
        }
        return "";
    }

    String displayLabel() {
        return displayLabel(errorCode, phase, status);
    }

    private static String humanizeCode(String code) {
        String[] parts = code.toLowerCase(Locale.US).split("_");
        StringBuilder sb = new StringBuilder();
        for (String part : parts) {
            if (part.isEmpty()) continue;
            if (sb.length() > 0) sb.append(' ');
            sb.append(Character.toUpperCase(part.charAt(0)));
            if (part.length() > 1) sb.append(part.substring(1));
        }
        return sb.toString();
    }

    private static String capitalize(String value) {
        if (value == null || value.isEmpty()) return "";
        return Character.toUpperCase(value.charAt(0)) + value.substring(1);
    }
}

File: go-sidecar/phase_result.go
package main

import (
	"fmt"
	"strings"
)

// PhaseResult is the v1 per-phase probe outcome carried on scan NDJSON results.
type PhaseResult struct {
	Phase      string         `json:"phase"`
	Status     string         `json:"status"`
	DurationMS int64          `json:"duration_ms"`
	ErrorCode  string         `json:"error_code,omitempty"`
	Retryable  bool           `json:"retryable"`
	Evidence   map[string]any `json:"evidence,omitempty"`
}

func newPhaseSuccess(phase string, durationMS int64) PhaseResult {
	return PhaseResult{
		Phase:      phase,
		Status:     "success",
		DurationMS: durationMS,
		Retryable:  false,
		Evidence:   map[string]any{},
	}
}

func newPhaseFailure(phase string, err error, durationMS int64, errorCode string) PhaseResult {
	code := strings.TrimSpace(errorCode)
	if code == "" && err != nil {
		code = classifyNetworkError(err, phase)
	}
	return PhaseResult{
		Phase:      phase,
		Status:     phaseStatusFromCode(code, err),
		DurationMS: durationMS,
		ErrorCode:  code,
		Retryable:  phaseRetryable(code),
		Evidence:   boundedPhaseEvidence(err),
	}
}

func phaseStatusFromCode(code string, err error) string {
	if code == "" && err == nil {
		return "success"
	}
	upper := strings.ToUpper(code)
	switch {
	case strings.HasSuffix(upper, "_TIMEOUT"):
		return "timeout"
	case strings.HasSuffix(upper, "_REFUSED"):
		return "refused"
	case strings.HasSuffix(upper, "_RESET"):
		return "reset"
	case strings.Contains(upper, "_SKIPPED"), strings.HasSuffix(upper, "_EXCLUDED"):
		return "skipped"
	case strings.Contains(upper, "_UNSUPPORTED"):
		return "unsupported"
	case strings.HasSuffix(upper, "_CANCELLED"):
		return "cancelled"
	default:
		return "failed"
	}
}

func phaseRetryable(code string) bool {
	upper := strings.ToUpper(strings.TrimSpace(code))
	switch {
	case strings.HasSuffix(upper, "_TIMEOUT"):
		return true
	case strings.HasSuffix(upper, "_RESET"):
		return true
	case upper == "DNS_TRUNCATED_TCP_RETRY_FAILED":
		return true
	default:
		return false
	}
}

func boundedPhaseEvidence(err error) map[string]any {
	if err == nil {
		return nil
	}
	detail := redactSensitiveString(err.Error())
	if len(detail) > 200 {
		detail = detail[:200]
	}
	return map[string]any{"detail": detail}
}

func appendTLSOutcomePhases(phases []PhaseResult, candidateSNI string, certVerified bool, elapsed int64) []PhaseResult {
	phases = append(phases, newPhaseSuccess("tcp", elapsed))
	if strings.TrimSpace(candidateSNI) != "" && !certVerified {
		return append(phases, newPhaseFailure("tls", fmt.Errorf("hostname verification failed"), elapsed, "TLS_VERIFY_HOSTNAME_MISMATCH"))
	}
	return append(phases, newPhaseSuccess("tls", elapsed))
}

func httpPhaseFromALPN(alpn string) string {
	if strings.Contains(strings.ToLower(alpn), "h2") {
		return "http2"
	}
	return "http1"
}

func finalizeFinalPhase(res result, phases []PhaseResult, lastErrCode string) string {
	if res.HTTP {
		return httpPhaseFromALPN(res.ALPN)
	}
	if res.TLS {
		return "tls"
	}
	if res.TCP {
		return "tcp"
	}
	for i := len(phases) - 1; i >= 0; i-- {
		if phases[i].Status != "success" && phases[i].Status != "skipped" {
			return phases[i].Phase
		}
	}
	if lastErrCode != "" {
		return "tcp"
	}
	return ""
}

func resultIndicatesTimeout(res result) bool {
	for _, phase := range res.PhaseResults {
		if phase.Status == "timeout" {
			return true
		}
	}
	return strings.HasSuffix(res.ErrorCode, "_TIMEOUT")
}

func resultIndicatesReset(res result) bool {
	for _, phase := range res.PhaseResults {
		if phase.Status == "reset" {
			return true
		}
	}
	return strings.HasSuffix(res.ErrorCode, "_RESET")
}
